README.md
# TakeOff15112TermProject
My term project for 15112 Fall 2020

Project Name: TakeOff: RC Airplane Simulator



Project Description:
	The goal of this RC airplane simulator is to help users learn to fly RC aircraft. In the RC hobby, simulators are essential 
tools, as they can help teach someone how to fly an aircraft before they fly it in real life and can either damage the aircraft or 
endanger others. I hope to have the user be able to plug in a controller, calibrate it, and then be able to fly the plane around 
an open world. There will also be a collection of activities that a user can use to practice specific skills to help them learn the 
basics of flying RC airplanes.



Running the project:
The "mainvX" Python file must be in the same folder as the "media" folder and "cmu_112_graphics_modified," as well as in the same folder as all
the other python files. To execute the program, you have to open the "mainvX" file and run the code in this file (Ctrl-B in VSCode)

Also make sure to use the cmu_112_graphics_modified.py file, which has lines 475, 483, 497, 510 and 556 commented out. 
This keeps keyPressed and mouse movements from calling redrawAll(), since in my case, timerFired() will constantly trigger redrawAll anyways.

Ensure that pygame 2.X is installed as well as Shapely. More info at bottom of this file


Using the code


Controller Setup:
There is no keyboard support, only controller support. To use a controller, plug it in and click "start"

-To calibrate a controller, open the "Controller" menu, hit "reset" and move all joysticks in a complete circle at the full 
	extent of their motion.
-To reassign channels, identify the proper channel by looking at the "Raw Data" output and then click the blue box under "reassign"
next to the channel that you want to reassign. Use backspace to delete the previous entry and then type the new channel. Click
off the box or press enter to save the new assignment.



Using the Main Game Mode:
-While on the field, you can use your controller to take off the aircraft and fly it around the world. 
-The settings menu, which is accessed by using the mouse to press the settings button, allows the user to customize the features in the
main game mode

Keyboard shortcuts in main game mode:
-"r": resets the aircraft
-"Space": toggles between a stationary camera and a camera following the plane



Training Mode:
-The buttons can be used to navigate to the desired training. Once a training is completed, the user is put back on the training selection screen.



Settings Mode

Physics:
	-Settings are self-explanatory
	-min values of 0 and max values of 99

Graphics
	-Daytime: Sets time to day if checked, otherwise nighttime
	-Cloud Count: sets the number of clouds 0-99      Reduce to increase performance
	-Bush Count: sets the number of bushes 0-9999     Reduce to increase performance
	-View Range: sets the maximum distance objects (except clouds) can be seen at  0-999      Reduce to improve performance
	-Sun Angle: Sets the angle of the sun in the sky. 20-90, below 30 not recommended
	-Camera Angle: Sets the angle at which the following-mode camera views the plane. Higher angle means more vertical
	-Camera Dist: Sets the approximate distance the camera follows the plane at 0-99
	-Use Shapely: toggle between using Shapely and not using Shapely     Enable to increase performance
		-Disclaimer: Not using Shapely will result in periodic crashes, especially on take-off
Data Readouts:
	-Display Data: Toggles if any of the features other than "Binoculars" are displayed
	-Altitude: Displays the current altitude of the plane when enabled
	-Velocity: Displays the curennt velocity of the plane when enabled
	-Attitude: Displays the current attitude (angle above the horizontal) when enabled
	-Distance: Displays the current distance from the plane to the origin when enabled
	-Framerate: Displays the current framerate when enabled
	-Binoculars: Shows a zoomed in view of the plane in the corner when enabled. Small performance reduction




Notes:
Graphics Offsets used
	-When offsetting text, buttons, etc., I often used multiples of 1/4, 1/8, 1/16, 1/32, and 1/64, which is why these numbers appear
	throughout the code. This greatly helped with lining different things up with each other

The question mark box on the controller screen opens an internet tab, at least in Windows
If you don't want the program to do this, just don't click the button




External Libraries Needed:
-Pygame Version 2.X: https://pypi.org/project/pygame/
-Shapely: https://pypi.org/project/Shapely/
-Webbrowser: https://docs.python.org/3/library/webbrowser.html (Should be installed already)




TPSubmission/changelog.txt

********preTP0

-Lots of rendering test files
-Controller interface created
-Combinded rendering code with controller interface


mainv1:
-Shapely is used to help in rendering. no physics

********TP0

mainv2: 
-made airplane into a class

mainv3: 
-Added basic physics to airplane. 
-Made getFaces and renderFaces much less repetitive by making use of more helper functions
-Organized functions in the GameMode class

mainv4
-Made physics independent of framerate. Makes the physics not dependent on the power of the computer
-Sun and randomly generated clouds (made a cloud class)

mainv5
-Made small modifications to the Button and TextBox class. 
-Also added a CheckBox subclass of button that allows the user to easily toggle variables
-Made sky color change based on status of a checkbox on the splash screen for testing of the CheckBox class
-Improved readability of getTValue()

mainv6
-altered how colors are used (They are now kept as tuples until they are being drawn. This allows for scaling the color based on the time of day)
-made ground and plane colors depend on time of day
-moved the daylight checkbox to the calibration screen temporarily so it can be changed without rerunning the program

********TP1

mainv7
-separated out into a few files
-altered ground loading. Generates a large dict and only shows close squares of grass.
-Created transition between following mode and stationary mode
-Stated InputBox class. Put a few input boxes on the splash screen temporarily

mainv8
-changes made to accommodate TrainingMode
-Most rendering functions for GameMode moves to separate file

mainv9
-Added TrainingMode. Constructed framework for it and created "trainingFunctions" to store the info for performing the various
training exercises offered
-Added button to toggle between TrainingMode and GameMode 
-Bushes, altered shadow function to support bushes, make sunPos dependent on an angle, added crashing + forced reset

mainv10
-fixed framerate bug when switching back from controller calibration (did not fix, fixed in v11)
-added control customization for roll rate, yaw rate, etc. and for deadzone
-removed test input boxes from splash screen
-altered drag function to make it proportional to velocity squared
-made the inputBoxVars "drag" and "throttleStrength" ready to be implemented on settings screen

mainv11
-Added "toolbox" to display some variables to the user
-Greatly expanded on the settings screen
-Altered calibration screen so the "reassign" buttons are now my custom user input boxes
-Added more training levels
-Adjusted how variables are re-initialized when the mode switches back to that game mode.
-Put my own shape intersection code at the bottom of the "renderingFunctions" file
-Altered some of the button creation code in calibration to make it less repetitive.
-removed "daytime" checkbox on calibration screen, moved to settings

********TP2

mainv11
-Added more training levels

mainv12
-Added "binoculars" and added a toggle setting for them
-Finished training levels
-Shapely "toggle" in settings
-Made bushes much more efficient
-Added balloons
-Added quit screen
-Started help screen
-Broke checkBoxes
-Minor changes to framework

mainv13
-Fixed checkBoxes
-Added background colors on the screens
-Fixed fontsize bugs
-Fixed camera stuttering problem
-Added filter to framerate to smooth out performance while still basing physics on framerate
-Added text to help screen
-Minor modifications to Bush and Balloon to make them share a helper function
-Added more comments
-Added balloon shadows + strings
-Added "distance" readout
-Added "?" on the help screen and a settings icon on the settings screen
-removed some try/excepts
-slight color adjustments and UI tweaks


mainv13.py
#This file contains the code that initializes app-wide variables and the game modes

from cmu_112_graphics_modified import *
from myClasses import *
from gameModes import *
import pygame



#Initializes the game controller channels
#Defaults to my Spektrum Dx6i but can be changed in controller calibration
def createChannels(app):
    app.channels = []
    throttle = ControllerChannel(app.inputBoxVars["throttleIndex"], "throttle")
    app.channels.append(throttle)
    yaw = ControllerChannel(app.inputBoxVars["yawIndex"], "yaw")
    yaw.isReversed = True
    app.channels.append(yaw)
    pitch = ControllerChannel(app.inputBoxVars["pitchIndex"], "pitch")
    app.channels.append(pitch)
    roll = ControllerChannel(app.inputBoxVars["rollIndex"], "roll")
    roll.isReversed = True
    app.channels.append(roll)
    app.channelAssignment = {}
    for channel in app.channels:
        app.channelAssignment[channel.name] = channel


#Creates the ModalApp and adds the main game modes to app
class TakeOff(ModalApp):
    def appStarted(app):
        #Initializes variables shared between modes
        app.initializeInputVariables()

        #Creates the game modes
        app.gameMode = GameMode()
        app.calibrationMode = CalibrationMode()
        app.splashScreenMode = SplashScreenMode()
        app.trainingMode = TrainingMode()
        app.settingsMode = SettingsMode()
        app.helpMode = HelpMode()
        app.quitConfirmationMode = QuitConfirmationMode()

        #Begins Pygame
        pygame.init()

        #Creates the controller channels
        createChannels(app)

        #Sets the settings used by the game modes
        app.timerDelay = 1
        app.aspectRatio = 16/9
    
        #Launches the splashScreen mode
        app.previousMode = app.splashScreenMode
        app.setActiveMode(app.splashScreenMode)
       
     
    #Initializes the variables that the user can interact with
    def initializeInputVariables(app):
        app.inputBoxVars = {}
        app.inputBoxVars["deadzone"] = 1
        app.inputBoxVars["rollRate"] = 50
        app.inputBoxVars["pitchRate"] = 50
        app.inputBoxVars["yawRate"] = 50
        app.inputBoxVars["drag"] = 50
        app.inputBoxVars["throttleStrength"] = 30
        app.inputBoxVars["stallSpeed"] = 12
        app.inputBoxVars["turningDrag"] = 50
        app.inputBoxVars["liftCoeff"] = 50

        app.inputBoxVars["isDaytime"] = True
        app.inputBoxVars["cloudCount"] = 20
        app.inputBoxVars["bushCount"] = 1000
        app.inputBoxVars["viewRange"] = 200
        app.inputBoxVars["sunAngle"] = 45
        app.inputBoxVars["cameraAngle"] = 15
        app.inputBoxVars["followDistance"] = 5
        app.inputBoxVars["useShapely"] = True

        app.inputBoxVars["throttleIndex"] = 1
        app.inputBoxVars["yawIndex"] = 0
        app.inputBoxVars["pitchIndex"] = 3
        app.inputBoxVars["rollIndex"] = 2

        app.inputBoxVars["displayData"] = True
        app.inputBoxVars["altitudeReading"] = True
        app.inputBoxVars["attitudeReading"] = True
        app.inputBoxVars["velocityReading"] = True
        app.inputBoxVars["distanceReading"] = True
        app.inputBoxVars["framerateReading"] = True
        app.inputBoxVars["binoculars"] = True



#Runs the game
TakeOff(width=1200, height=675)


